### DISCUSSION TITLE
Properly change contrast to better expose text in cropped image from video capture

### DISCUSSION THREAD

**Discussion by qecnebu** [ImageMagick/ImageMagick#6296] [Help]
### ImageMagick version

6.9.11-60

### Operating system, version and so on

macOS Ventura 13.3.1

### Description

I have surveillance footage that has date/time information in the video that I am trying to extract.

I use `ffmpeg` to extract a frame three seconds into the video since the time stamp doesn't reliably appear right away:
```
ffmpeg -ss 00:00:03.00 -i video_file.MP3 -vframes 1 -f image2 video_file.jpg
```
I then use `convert` to crop out the section of the captured image to isolate the date/time text:
```
convert video_file.jpg -crop 455x45+1420+35 video_file_out.jpg
```

This works well enough to grab the information I need and I am then able to feed this info into `tesseract` to get the date/time info into a file for further modification.

This works without much fuss on one of my cameras, but the other cameras have problematic backgrounds that cause the ocr process to fail.

I've tried resizing and upping the contrast to accommodate for this, but for certain times of the day, the background is just too much and bleeds through with the text making it unreadable.

A original _out file that works pretty well for the ocr looks something like the below:
![image](https://user-images.githubusercontent.com/115429133/236337929-676be711-b1c6-4e4d-8686-3c6aac6de610.png)

I then run it through another `convert` pass to tweak the contrast:
```
convert video_file_out.jpg -resize 200% -brightness-contrast -40x100 -enhance -equalize -contrast video_file_resize.jpg
```

And get the following that works great for ocr:
![image](https://user-images.githubusercontent.com/115429133/236338222-69d55ff9-0566-4e0f-8a93-5735a9c38664.png)

However, sometimes the background bleeds into the date/time text in the original crop:
![image](https://user-images.githubusercontent.com/115429133/236338349-0553a90b-4536-4aa3-b32c-110192b29e0e.png)

And when I run this through the above secondary `convert`, I get something that just doesn't work for ocr:
![image](https://user-images.githubusercontent.com/115429133/236338467-e9206344-0783-4e68-9fe6-36a5778cfb7d.png)

I would ideally like to make the white text pop so it is very readable and the background be diminished so that it doesn't interfere with the `tesseract` ocr process.

I tried looking around some of the other posts, but I don't see anything that quite matches what I'm running into.

Any advice or pointers on what I should be doing would be greatly appreciated.

**Comment by qecnebu** [comment]
I stumbled upon the `-negate` and `-monochrome` flags, and those seem to help some, but there's just so much other noise in the one picture that things aren't being interpreted correctly.

I ended up with:
```
convert video_file.jpg  -resize 200% -brightness-contrast -10x100 -contrast -negate -monochrome video_file_resize.jpg
```

With the image without much of anything in the background, the above works great:
![image](https://user-images.githubusercontent.com/115429133/236482455-866bb95a-d4f6-4f3d-bd68-7decf38a3656.png)

For the other image with background noise, it makes for a little more legible date/time, but the other artifacts still get in the way:
![image](https://user-images.githubusercontent.com/115429133/236482946-c4fa1ece-2b6a-4cf8-8eec-724d2918392c.png)

I ended up removing the `-enhance -equalize -contrast` since it didn't really seem to help any.

I'm not sure what the best options would be to clean this image up enough to obscure the background while leaving the text in place and make it more legible.  Ironically, Ventura is able to OCR the info just fine as when I'm trying to select the image, it's trying to correctly select the text it "sees", which is pretty close what is in the image.

Again, any pointers on how to obscure the background so that only the date/time text are the most prominent thing displayed would be appreciated.

**Reply by fmw42** [reply]
Background removal is very hard to do with non-constant color backgrounds in 2D image processing.  I would suggest you look for some AI/Deep Learning based solution.  One that you can try is http://remove.bg

However, if your "noise" is only very small compared to your text characters and does not touch the text characters, then you could remove it using -connected-components.

**Comment by qecnebu** [comment]
Thanks, but I'm trying to script this to get the output of these image snippets in as automated a way as I can, so having to go online for something [which looks like you have to pay for anyways] isn't something that would work nor willing to do.  I did look into the -connect-components option mentioned, but it's very complex to try to implement.

Since the OCR seemed to be working more when I viewed these files through Preview, I looked to see if there was an option and there is something someone created that will leverage the new I believe Livetext functionality through the CLI.

I found `litex` at https://iosexample.com/live-text-command-line-tool-for-macos/

It is leaps and bounds better than `tesseract` and is able to work more reliably with the files I'm generating, which is a slight variation of the last thing I came up with:
```
convert video_file_out.jpg -brightness-contrast -70x100 -contrast -monochrome -negate -enhance -resize 150% video_file_resize.jpg
```

The converted text results still aren't 100%, but like I said earlier, they are so much better than `tesseract`.

I do imagine cleaning up these image further would help with this conversion process, but this will have to do.

**Reply by fmw42** [reply]
I see no reason to use both -brightness-contrast and -contrast.  You should be able to drop -contrast and increase the contrast in -brightness-contrast.
